### PR TITLE
HOTT-3738: Load rates from XE API once a month

### PR DESCRIPTION
**Pull request by maverickvi**
### Jira link

https://transformuk.atlassian.net/browse/HOTT-3738

### What?

I have added/removed/altered:

- [ ] Load rates from XE API once a month

### Why?

I am doing this because:

- We need to display these rates on the frontend

### Have you? (optional)

- [ ] Added new environment variables with correct values to all apps in all spaces

### CODE REVIEW

File: .env.development [hunk 1]
```diff
@@ -19,3 +19,6 @@ STEMMING_EXCLUSION_REFERENCE_ANALYZER=/usr/share/opensearch/config/stemming_excl
 SYNONYM_REFERENCE_ANALYZER=/usr/share/opensearch/config/synonyms_all.txt
 TARIFF_IGNORE_PRESENCE_ERRORS=1
 TARIFF_QUERY_SEARCH_PARSER_URL=http://127.0.0.1:5000/api/search/
+XE_API_USERNAME=user
+XE_API_PASSWORD=pass
+XE_API_URL=url
```
Comment (willfish) [inline .env.development:24]: Missing new line :scream_cat:  :laughing:

File: app/models/exchange_rates/xe_api.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    def get_all_historic_rates
+      send_request("/v1/historic_rate.json/?from=GBP&date=#{@date}&to=*&amount=1")
+    end
+
+    private
+
+    def send_request(url)
+      uri = URI(TradeTariffBackend.xe_api_url + url)
+
+      client = Faraday.new(uri.host) do |conn|
```
Comment (willfish) [inline app/models/exchange_rates/xe_api.rb:18]: How do you feel about memoising this client in the class singleton?

File: spec/models/exchange_rates/xe_api_spec.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        },
+      ],
+    }.to_json
+  end
+
+  let(:faraday_mock) { instance_double(Faraday::Connection) }
+
+  before do
+    allow(Faraday).to receive(:new).and_return(faraday_mock)
+    allow(TradeTariffBackend).to receive(:xe_api_url).and_return('xeurl.com')
```
Comment (willfish) [inline spec/models/exchange_rates/xe_api_spec.rb:28]: I guess you can control this via the .env.test file?

File: spec/services/exchange_rates/monthly_exchange_rates_service_spec.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    end
+
+    # rubocop:disable all
+    context 'when tomorrow is penultimate Thursday and today is Wednesday' do
+      before do
+        travel_to Time.zone.local(2023, 7, 19)
+      end
+
+      it 'calls the necessary services and notifies' do
+        expect(ExchangeRates::UpdateCurrencyRatesService).to receive_message_chain(:new, :call)
```
Comment (willfish) [inline spec/services/exchange_rates/monthly_exchange_rates_service_spec.rb:18]: I think receive_message_chain is frowned on?

File: spec/services/exchange_rates/monthly_exchange_rates_service_spec.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+RSpec.describe ExchangeRates::MonthlyExchangeRatesService do
+  describe '.call' do
+    subject(:call_service) { described_class.call }
+
+    after do
+      travel_back
+    end
+
+    # rubocop:disable all
```
Comment (willfish) [inline spec/services/exchange_rates/monthly_exchange_rates_service_spec.rb:11]: This is a bit of a sledge hammer. I'd just not use

File: spec/services/exchange_rates/update_currency_rates_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+          expect {
+            service.call
+          }.to change(ExchangeRateCurrencyRate, :count).by(1)
+        end
+
+        # rubocop:disable RSpec/MultipleExpectations
+        it 'raises unique index error when trying to create duplicate rates' do
+          create(:exchange_rate_currency_rate)
+
+          expect {
```
Comment (willfish) [inline spec/services/exchange_rates/update_currency_rates_service_spec.rb:42]: This smells a bit. Can't the application just handle these gracefully and you check the effect? So I'd opt for an upsert which can either update the rows you want (say just the audit timestamps) or do nothing

File: app/models/exchange_rates/xe_api.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+module ExchangeRates
+  class XeApi
+    class XeApiError < StandardError; end
+
+    def initialize(date: Time.zone.today)
+      @date = date.strftime('%Y-%m-%d')
+    end
+
+    def get_all_historic_rates
+      send_request("/v1/historic_rate.json/?from=GBP&date=#{@date}&to=*&amount=1")
```
Comment (alexdesi) [inline app/models/exchange_rates/xe_api.rb:10]: it's a good practice to put  the endpoint `/v1/historic_rate.json/ ` in a constant

File: app/models/exchange_rates/xe_api.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    def send_request(url)
+      uri = URI(TradeTariffBackend.xe_api_url + url)
+
+      client = Faraday.new(uri.host) do |conn|
+        conn.request :basic_auth, TradeTariffBackend.xe_api_username, TradeTariffBackend.xe_api_password
+      end
+
+      response = client.get(uri)
+
+      if response.success?
```
Comment (alexdesi) [inline app/models/exchange_rates/xe_api.rb:24]: If I am not wrong, Faraday does raise exceptions if the HTTP response is not a success (response outside 200..299 range). So I think we don't need manual check for success.

File: app/services/exchange_rates/monthly_exchange_rates_service.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+      def notify
+        message = 'Exchange rates for the current month have been added and are accessible for viewing at /exchange_rates.'
+
+        logger.info message
+
+        SlackNotifierService.call(message)
+      end
+
+      def tomorrow_is_penultimate_thursday?
```
Comment (alexdesi) [inline app/services/exchange_rates/monthly_exchange_rates_service.rb:24]: Should we name this with the convention `is_`at the start?
`is_tomorrow_penultimate_thursday?` or
`is_tomorrow_the_penultimate_thursday?`

File: app/models/exchange_rates/xe_api.rb (hunk 8 above)
Comment (willfish) [reply app/models/exchange_rates/xe_api.rb:24]: I think this depends on middleware

Comment (willfish) [reply app/models/exchange_rates/xe_api.rb:24]: https://github.com/trade-tariff/trade-tariff-frontend/blob/main/app/services/client_builder.rb#L52

**Review by willfish — APPROVED** [review]

**Pull request merged**